### PR TITLE
[Tests] Add more cases to sighash_tests for malleated txes

### PR DESCRIPTION
**Pull request by random-zebra**
Create random txes, then add/modify/remove inputs, outputs, sapling data fields, and check the hash of the data serialized for transparent signatures.

### CODE REVIEW

**Comment by furszy** [comment]
Moving the recent PM convo result here.
 Let's keep the `sapling_malleable_sigs.py` for now. It's helpful to verify the sapling transaction parsing (not really its purpose but.. helpful at the end). We can remove it once we get more coverage on the area.

**Comment by random-zebra** [comment]
:+1: dropped last commit.

**Review by Fuzzbawls — APPROVED** [review]
ACK 76cea0bcba279a8d7c3a0bea527dcb7dcdd63e3f

File: src/test/sighash_tests.cpp [hunk 1]
```diff
@@ -244,6 +244,256 @@ BOOST_AUTO_TEST_CASE(sighash_from_data)
...
+        for (int nIn = 0; nIn < (int) tx.vin.size(); nIn++) {
+            BOOST_CHECK(vsh[nIn] == SignatureHash(vScriptCode[nIn], CTransaction(mtx), nIn, SIGHASH_ALL, 0, tx.GetRequiredSigVersion()));
+        }
+
+        // remove one transparent input
+        if (mtx.vin.size() > 0) {
+            mtx.vin.pop_back();
+            CTransaction tx2(mtx);
+            for (int nIn = 0; nIn < (int) tx.vin.size() - 1; nIn++) {
+                BOOST_CHECK(vsh[nIn] != SignatureHash(vScriptCode[nIn], tx2, nIn, SIGHASH_ALL, 0, tx2.GetRequiredSigVersion()));
```
Comment (furszy) [inline src/test/sighash_tests.cpp:280]: would be good to use `BOOST_CHECK_MESSAGE` to be more precise on the error description (same for the rest).

File: src/test/sighash_tests.cpp [hunk 2]
```diff
@@ -244,6 +244,256 @@ BOOST_AUTO_TEST_CASE(sighash_from_data)
...
+            for (int nIn = 0; nIn < (int) tx.vin.size(); nIn++) {
+                BOOST_CHECK(vsh[nIn] != SignatureHash(vScriptCode[nIn], CTransaction(mtx), nIn, SIGHASH_ALL, 0, SigVersion::SIGVERSION_SAPLING));
+            }
+
+            // modify one spend (cv, anchor, nullifier, rk, zkproof, spendAuthSig)
+            int i = InsecureRandRange(tx.sapData->vShieldedSpend.size());
+
+            mtx = _tx;
+            while (mtx.sapData->vShieldedSpend[i].cv == tx.sapData->vShieldedSpend[i].cv)
+                mtx.sapData->vShieldedSpend[i].cv = GetRandHash();
```
Comment (furszy) [inline src/test/sighash_tests.cpp:320]: I'm interested to know why did you added this, it's an overkill for something that should never happen.

**Review by furszy — COMMENTED** [review]
Initial review, looking pretty nice, cool work 👌 .

File: src/test/sighash_tests.cpp (hunk 1 above)
Comment (random-zebra) [reply src/test/sighash_tests.cpp:280]: I think it's fine. They are all checking the same thing.

File: src/test/sighash_tests.cpp (hunk 2 above)
Comment (random-zebra) [reply src/test/sighash_tests.cpp:320]: Same as all the other checks.

**Review by furszy — APPROVED** [review]
Really nice 👌, this test coverage made my soul happy.
ACK 76cea0bcba279a8d7c3a0bea527dcb7dcdd63e3f

**Pull request merged**